background_scripts/completion.coffee

```diff
@@ -321,27 +328,80 @@ class TabCompleter
+    if custom
+      query = queryTerms[1..].join " "
+      title = if haveDescription then query else keyword + ": " + query
+      suggestions.push @mkSuggestion null, queryTerms, type, mkUrl(query), title, @computeRelevancy, 1
+      suggestions[0].autoSelect = true
+      queryTerms = queryTerms[1..]
+    else
+      query = queryTerms.join " "
+
+    if queryTerms.length == 0
+      return onComplete suggestions
+
+    onComplete suggestions, (existingSuggestions, onComplete) =>
+      suggestions = []
+      # For custom search-engine queries, this adds suggestions only if we have a completer.  For other queries,
+      # this adds suggestions for the default search engine (if we have a completer for that).
+
+      # Scoring:
+      #   - The score does not depend upon the actual suggestion (so, it does not depend upon word
+      #     relevancy).  We assume that the completion engine has already factored that in.  Also, completion
+      #     engines often handle spelling mistakes, in which case we wouldn't find the query terms in the
+      #     suggestion anyway.
+      #   - The score is higher if the query term is longer.  The idea is that search suggestions are more
+      #     likely to be relevant if, after typing some number of characters, the user hasn't yet found
+      #     a useful suggestion from another completer.
+      #   - Scores are weighted such that they retain the order provided by the completion engine.
+      characterCount = query.length - queryTerms.length + 1
+      score = 0.6 * (Math.min(characterCount, 10.0)/10.0)
+
+      if 0 < existingSuggestions.length
+        existingSuggestionMinScore = existingSuggestions[existingSuggestions.length-1].relevancy
+        if score < existingSuggestionMinScore and MultiCompleter.maxResults <= existingSuggestions.length
+          # No suggestion we propose will have a high enough score to beat the existing suggestions, so bail
+          # immediately.
+          return onComplete []
+
+      SearchEngines.complete searchUrl, queryTerms, (searchSuggestions = []) =>
+        for suggestion in searchSuggestions
+          insertText = if custom then "#{keyword} #{suggestion}" else suggestion
+          suggestions.push @mkSuggestion insertText, queryTerms, type, mkUrl(suggestion), suggestion, @computeRelevancy, score
+          score *= 0.9
+
+        # We keep at least three suggestions (if possible) and at most six.  We keep more than three only if
+        # there are enough slots.  The idea is that these suggestions shouldn't wholly displace suggestions
+        # from other completers.  That would potentially be a problem because there is no relationship
+        # between the relevancy scores produced here and those produced by other completers.
+        count = Math.min 6, Math.max 3, MultiCompleter.maxResults - existingSuggestions.length
+        onComplete suggestions[...count]
+
+  mkSuggestion: (insertText, args...) ->
```

Comment:
This (and the `Suggestion` constructor) takes quite a lot of arguments. Maybe accepting a key-value pair would make it easier to check that the right arguments have been set, and ensure the order isn't wrong? A list of the arguments in a comment here would also be helpful.